File: 2023/day_03/js/gear_ratios.js
import fs from "node:fs";
import { isNumber } from "node:util";

const readFile = (path) => {
    fs.readFile(path, 'utf8', (err, data) => {
        if (err) {
            console.log(err);
            throw err;
        }
        // rest of code
        const arr = makeArray(data);
        arr.pop();
        const result = sumOf(arr);
        console.log("Result of part one: " + result);
    });
};

const makeArray = (data) => {
    const arr = data.split('\n');
    return arr;
};

const sumOf = (arr) => {
    let count = 0;
    const nums = [];
    // console.log("Length of string is " + arr[0].length);
    for (let i = 0; i < arr.length; i++) {
        // for (let i = 2; i < 3; i++) {
        const num = [];
        let temp = '';
        const indexes = {};
        for (let j = 0; j < arr[i].length; j++) {
            const current = arr[i][j];
            const next = arr[i][j + 1];
            const third = arr[i][j + 2];
            // const prev = arr[i][j-1];
            if (!isDigit(current) || isSymbol(current)) {
                console.log("Dot");
                // continue;
            }

            // one digit
            if ((arr[i][j-1] === undefined || !isDigit(arr[i][j-1])) && isDigit(current) && indexes[j] === undefined) {
                indexes[j] = arr[i][j];
                temp += current;
                num.push(j);
            }

            // two digit
            if (isDigit(next) && indexes[j + 1] === undefined) {
                num.push(j + 1);
                indexes[j + 1] = arr[i][j + 1];
                temp += arr[i][j + 1];
            }

            // third
            if (isDigit(third) && indexes[j + 2] === undefined) {
                num.push(j + 2);
                indexes[j + 2] = arr[i][j + 2];
                temp += arr[i][j + 2];
            }
            console.log("Checking for " + current);

            // Checking numbers
            if (nums[nums.length - 1] != temp && temp != "" && temp.length <= 3) {
                const check = isValid(arr, i, num);
                nums.push(temp);
                if (check !== null) {
                    console.log("Temp: " + temp);
                    count += Number(temp);
                }
            }
        }
    }
    return count;
}

const isValid = (arr, line, ind) => {
    console.log("Line: " + line + " => Indexes: " + ind);
    for (let i = 0; i < ind.length; i++) {
        const current = ind[i];
        const next = ind[i] + 1;
        const prev = ind[i] - 1;
        // console.log("Coordinates: (" + line + "," + current + ")");
        // horizontal right
        if (arr[line][current + 1] !== undefined && !isDigit(arr[line][next]) && isSymbol(arr[line][next])) {
            console.log("Right valid! (" + line + "," + (current + 1) + ")");
            return 1;
        }
        // horizontal left
        if (arr[line][current - 1] !== undefined && !isDigit(arr[line][prev]) && isSymbol(arr[line][prev])) {
            console.log("Left valid! (" + line + "," + (current - 1) + ")");
            return 1;
        }
        // vertical down
        if (arr[line + 1] !== undefined && !isDigit(arr[line+1][current]) && isSymbol(arr[line+1][current])) {
            console.log("Down valid! (" + (line + 1) + "," + (current) + ")");
            return 1;
        }
        // vertical up
        if (arr[line - 1] !== undefined && !isDigit(arr[line - 1][current]) && isSymbol(arr[line - 1][current])) {
            console.log("Up valid! (" + (line - 1) + "," + (current) + ")");
            return 1;
        }
        // diagonal to bottom right
        if (arr[line+1] !== undefined && arr[line + 1][next] && current !== arr.length - 1 && !isDigit(arr[line+1][next]) && isSymbol(arr[line+1][next])) {
            console.log("Down valid! (" + (line + 1) + "," + (current) + ")");
            return 1;
        }
        // diagonal to bottom left
        if (arr[line+1] !== undefined && arr[line + 1][prev] !== undefined && current !== 0 && !isDigit(arr[line+1][prev]) && isSymbol(arr[line+1][prev])) {
            console.log("bottom left");
            return 1;
        }
        // diagonal to upper right
        if (arr[line-1] !== undefined && current !== arr.length - 1 && !isDigit(arr[line-1][next]) && isSymbol(arr[line-1][next])) {
            console.log("upper right");
            return 1;
        }
        // diagonal to upper left
        if (arr[line-1] !== undefined && current !== 0 && !isDigit(arr[line-1][prev]) && isSymbol(arr[line-1][prev])) {
            console.log("upper left");
            return 1;
        }
    }
    return null;
};

const isDigit = (chr) => {
    const number = /^\d+$/;
    if (number.test(chr)) {
        return true;
    }
    return false;
};

const isSymbol = (char) => {
    const dot = /^[!@#\$%\^\&*\)\(+=_-]+$/;
    if (char.match(dot)) {
        return true;
    }

    return false;
}

const main = (file) => {
    console.log("Advent of Code 2023\nDay 03: Gear Ratios");

    readFile(file);
};

// main("../test.txt");
main("../input.txt");
console.log("Wrong answer 1: 514400");
console.log("Wrong answer 2: 514455");

File: 2023/day_03/lua/gear_ratios.lua
-- = = = = = = = = = = = == = = --
--   Advent of Code || Day 03   --
-- = = = = = = = = = = = == = = --

package.path = package.path .. ";../../../modules/modules.lua"
local aoc = require("modules")

-- local file = aoc.fileRead('../test.txt')
-- local file = aoc.fileRead('../test1.txt')
local file = aoc.fileRead('../input.txt')

local lines = aoc.explode("\n", file)

local function isSymbol(char)
    local dot = "^[!@#%$%^&*()%+=_-]+$"
    if char ~= nil and string.match(char, dot) then
        return true
    end
    return false
end


local function isValid(array, i, index)
    -- print("Line: "..i.." Indexes: " .. table.concat(index, ", "))
    for _, ind in ipairs(index) do
        -- local current = array[i]:sub(ind, ind)
        local next = array[i]:sub(ind + 1, ind + 1)
        local prev = array[i]:sub(ind - 1, ind - 1)
        local up, down, bottomRight, bottomLeft, upperRight, upperLeft

        if i - 1 >= 1 then
            up = array[i - 1]:sub(ind, ind)
            if ind - 1 >= 1 then
                upperLeft = array[i - 1]:sub(ind - 1, ind - 1)
            end
            if ind + 1 <= #array[ind] then
                upperRight = array[i - 1]:sub(ind + 1, ind + 1)
            end
        end
        if i + 1 <= #array then
            down = array[i + 1]:sub(ind, ind)
            if ind - 1 >= 1 then
                bottomLeft = array[i + 1]:sub(ind - 1, ind - 1)
            end
            if ind + 1 <= #array[i] then
                bottomRight = array[i + 1]:sub(ind + 1, ind + 1)
            end
        end

        -- right
        if next and not tonumber(next) and isSymbol(next) then
            -- print("Found right!")
            return true
        end
        -- left
        if prev and not tonumber(prev) and isSymbol(prev) then
            -- print("Found left!")
            return true
        end
        -- up
        if up and not tonumber(up) and isSymbol(up) then
            -- print("Found up!")
            return true
        end
        -- down
        if down and not tonumber(down) and isSymbol(down) then
            -- print("Found down!")
            return true
        end
        -- bottom right
        if bottomRight and not tonumber(bottomRight) and isSymbol(bottomRight) then
            -- print("Found BOTTOM right!")
            return true
        end
        -- bottom left
        if bottomLeft and not tonumber(bottomLeft) and isSymbol(bottomLeft) then
            -- print("Found BOTTOM left!")
            return true
        end
        -- upper right
        if upperRight and not tonumber(upperRight) and isSymbol(upperRight) then
            -- print("Found BOTTOM right!")
            return true
        end
        -- upper left
        if upperLeft and not tonumber(upperLeft) and isSymbol(upperLeft) then
            -- print("Found BOTTOM left! Char: " .. upperLeft)
            return true
        end
    end
    return false
end

local function findNumber(array)
    local count = 0
    local nums = {}
    for i = 1, #array do
        local checker = {}
        local temp = ""
        local indexes = {}
        for j = 1, #array[i] do
            local char = array[i]:sub(j, j)
            -- print("Char: " .. char)
            if tonumber(char, 10) then
                -- one digit
                if temp == "" and not checker[j] then
                    temp = char
                    checker[j] = true
                    table.insert(indexes, j)
                end
                local next = array[i]:sub(j + 1, j + 1)
                if not tonumber(next, 10) then
                    -- print("One digit")
                    table.insert(nums, temp)
                    local valid = isValid(array, i, indexes)
                    if valid then
                        table.insert(nums, temp)
                        count = count + tonumber(temp)
                    end
                    temp = ""
                    indexes = {}
                end

                -- two digit
                if tonumber(next, 10) and not checker[j + 1] then
                    checker[j + 1] = true
                    temp = temp .. next
                    table.insert(indexes, j + 1)
                end
                local third = array[i]:sub(j + 2, j + 2)
                if tonumber(third, 10) == nil and #indexes ~= 0 then
                    local valid = isValid(array, i, indexes)
                    if valid then
                        table.insert(nums, temp)
                        count = count + tonumber(temp)
                    end
                    temp = ""
                    indexes = {}
                end

                -- three digit
                if tonumber(third, 10) and not checker[j + 2] then
                    checker[j + 2] = true
                    temp = temp .. third
                    table.insert(indexes, j + 2)
                end

                local forth = array[i]:sub(j + 3, j + 3)
                if not tonumber(forth, 10) then
                    table.insert(nums, temp)
                    local valid = isValid(array, i, indexes)
                    if valid then
                        table.insert(nums, temp)
                        count = count + tonumber(temp)
                    end
                    temp = ""
                    indexes = {}
                end
            end
        end
        indexes = {}
    end
    -- print(table.concat(nums, " "))
    return count
end

local result = findNumber(lines)
print("Result of part one: " .. result)
